doc/neps/nep-0051-scalar-representation.rst

```diff
@@ -0,0 +1,332 @@
+the largest changes needed in the test suite.
+
+The proposed changes for void scalars and masked ``fill_value`` makes it
+necessary to expose the scalar representation without the type.
+
+We propose introducing the semi-public API::
+
+    np.lib.arrayprint.get_formatter(*,
+            data=None, dtype=None, fmt=None, options=None)
+
+to replace the current internal ``_get_formatting_func``.  This will allow
+two things compared to the old function:
+* ``data`` may be ``None`` (if ``dtype`` is passed) allowing to not pass
+  multiple values that will be printed/formatted later.
+* ``fmt=`` will allow passing on format strings to a DType specific element
+  formatter in the future.  For now, ``get_formatter()`` it will accept
+  ``repr`` or ``str`` (the singletons not strings) to format the elements
+  without type information (``'3.1'`` rather than ``np.longdouble('3.1')``).
+  The implementation ensures that formatting matches except for the type
+  information.
+
+  The empty format string will print identically to ``str()`` (with possibly
+  extra padding when data is passed).
+
+``get_formatter()`` is expected to query a user DTypes method in the future
+allowing customized formatting for all DTypes.
+
+Making it public allows the use for ``np.record`` and masked arrays.
+Currenlty, the formatters themselves seem semi-public and using a single
+entry-point will hopefully provide a clear API for formatting NumPy values.
+
+The large part for the scalar representation changes had previously been done
+by Ganesh Kathiresan in [2]_.
+
+Alternatives
+============
+
+Different representation could be discussed, main alternatives are spelling
+``np.`` as ``numpy.`` or dropping the ``np.`` part from the numerical scalars.
+We believe that using ``np.`` is sufficiently clear, concise, and does allow
+copy pasting the representation.
+Using only ``float64(3.0)`` without the ``np.`` prefix is more concise but
+contexts may exists where the NumPy dependency is not fully clear and the name
+could clash with other libraries.
+of ``numpy`` or ``np`` for the numerical types to give for example
+``np.float64(3.0)``.
+
+For booleans an alternative would be to use ``np.bool_(True)`` or ``bool_(True)``.
+However, NumPy boolean scalars are singletons and the proposed formatting is more
+concise.  Alternatives for booleans were also discussed previously in [1]_.
+
+For the string scalars, the confusion is generally less pronounced.  It may be
+reasonable to defer changing these.
+
+Non-finite values
+-----------------
+The proposal does not allow copy pasting ``nan`` and ``inf`` values.
+They could be represented by ``np.float64('nan')`` or ``np.float64(np.nan)``
+instead.
+This is more concise and Python also uses ``nan`` and ``inf`` rather than
+allowing copy-pasting by showing it as ``float('nan')``.  Arguably, it would be
+a smaller addition in NumPy, where the will already be always printed.
+
+``get_formatter()``
```

Comment:
It seems like the point of this section is to highlight the differences between array and scalar formatting - maybe a more detailed title or an example would help get the point across (assuming I've understood correctly...)

Reply:
Oh, the point is that `get_formatter()` introduction has some choices to be made and I wanted to list them.  We need something.  I changed the title to hopefully make that clear.